### PR TITLE
[DNM] Arch out-of-tree

### PR DESCRIPTION
**Pull request by kapedev**
(This is based on-top of PR #11309, and should not be merged)
Proposal for the support of having a processor architecture out-of-tree.
It includes moving files from the two following locations:
${ZEPHYR_BASE}/arch/
${ZEPHYR_BASE/include/arch
into an out-of-tree path with the same tree structure

This is done by supplying the optional argument ARCH_ROOT, in the same way it is done with BOARD_ROOT and SOC_ROOT.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -49,13 +49,14 @@ zephyr_library_named(zephyr)
...
   kernel/include
-  arch/${ARCH}/include
+  ${ARCH_DIR}/${ARCH}/include
   ${SOC_DIR}/${ARCH}/${SOC_PATH}
   ${SOC_DIR}/${ARCH}/${SOC_PATH}/include
   ${SOC_DIR}/${ARCH}/${SOC_FAMILY}/include
   ${BOARD_DIR}
   include
   include/drivers
+  ${ARCH_INCL_DIR}
```
Comment (SebastianBoe) [inline CMakeLists.txt]: I'm not seeing the need for ARCH_INCL_DIR.

Would it be possible to run

zephyr_include_directories(..)

from within the out-of-tree ARCH ?

Comment (kapedev) [reply CMakeLists.txt]: You are right that the same can be achieved by adding the following line:
`zephyr_include_directories(../../include)` to the out-of-tree file {ARCH_ROOT}/arch/{ARCH}/CMakeLists.txt

The trade-off beeing that this then requires arch's that are moved out-of-tree to add that line. With the ${ARCH_INCL_DIR} it does not. Both solutions will work.

I will update the PR with your suggestions

**Comment by kapedev** [comment]
Failing due to Cmake version beeing 3.8.2, still awaiting a decision on bumping this. Works on cmake 3.9.0

**Comment by mped-oticon** [comment]
I guess we can resume this now that we have bumped cmake up?


On a side note: I think it would be good to do once-forever solution to the out-of-tree, so we do not have do an out-of-tree for board, arch, toolchain, and what may come in future etc.
I think a viable way for that would be to exclusively use the zephyr versions of cmake commands in `extensions.cmake`. Then we could have a zephyr_add_subdirectory to _also_ look for the same dir, relatively, under another zephyr root. This would effectively give the union of two trees.

**Comment by SebastianBoe** [comment]
> Then we could have a zephyr_add_subdirectory to _also_ look for the same dir, relatively, under another zephyr root. This would effectively give the union of two trees.

This sounds promising. This approach has not been considered before.

**Comment by kapedev** [comment]
This has been superseded by PR #13130

**Comment by kapedev** [comment]
> I guess we can resume this now that we have bumped cmake up?
> 
> On a side note: I think it would be good to do once-forever solution to the out-of-tree, so we do not have do an out-of-tree for board, arch, toolchain, and what may come in future etc.
> I think a viable way for that would be to exclusively use the zephyr versions of cmake commands in `extensions.cmake`. Then we could have a zephyr_add_subdirectory to _also_ look for the same dir, relatively, under another zephyr root. This would effectively give the union of two trees.

Agreed. Eventually we will also need a Subsystem/Driver-out-of-tree and we should investigate this approach

**Pull request closed without merging**